File: src/prompt_manager/static/js/modules/conversation_mode.js
/**
 * Conversation Mode Module
 *
 * Manages the conversation mode state machine and UI interactions.
 * Implements the State Pattern for voice conversation flow.
 *
 * Two Modes:
 * - Manual Mode: IDLE → LISTENING (immediate transcription)
 * - Hands-Free Mode: IDLE → WAKE_LISTENING (waits for wake word)
 *
 * State Transitions (Manual Mode):
 * - IDLE → LISTENING (activate)
 * - LISTENING → PAUSED (pause)
 * - LISTENING → SENDING (send message)
 * - PAUSED → LISTENING (resume)
 * - SENDING → PLAYING (receive response)
 * - PLAYING → LISTENING (finish playback)
 * - ANY → IDLE (deactivate)
 *
 * State Transitions (Hands-Free Mode):
 * - IDLE → WAKE_LISTENING (activate with hands-free enabled)
 * - WAKE_LISTENING → LISTENING (wake word detected: "Hey Amber")
 * - LISTENING → WAKE_LISTENING (sleep word detected: "Sleep Amber")
 * - LISTENING → SENDING (10 seconds of silence triggers auto-send)
 * - SENDING → PLAYING (receive response)
 * - PLAYING → LISTENING (finish playback, ready for next question)
 * - ANY → IDLE (deactivate)
 *
 * Dependencies:
 * - VoiceInteraction module for controlling mic/speech
 * - showNotification function from main dashboard
 * - StateIndicator for visual state feedback
 */

import { WakeWordDetector } from './wake_word_detector.js';
import { SilenceDetector } from './silence_detector.js';

// Detector instances for hands-free mode
const wakeWordDetector = new WakeWordDetector();
const silenceDetector = new SilenceDetector();

/**
 * Helper to update state and notify state indicator.
 */
function updateState(newState) {
    conversationMode.state = newState;
    if (StateIndicator && StateIndicator.updateState) {
        StateIndicator.updateState(newState);
    }
}

// Conversation mode state machine (JavaScript implementation of domain model)
const conversationMode = {
    state: 'IDLE',
    isActive: false,
    handsFreeModeEnabled: false,

    enableHandsFreeMode() {
        this.handsFreeModeEnabled = true;
    },

    disableHandsFreeMode() {
        this.handsFreeModeEnabled = false;
    },

    onWakeWordDetected() {
        if (this.state === 'WAKE_LISTENING') {
            updateState('LISTENING');
        } else {
            console.warn('Wake word detected but not in WAKE_LISTENING state');
        }
    },

    onSleepWordDetected() {
        if (!this.handsFreeModeEnabled) {
            return;
        }
        if (this.state === 'LISTENING') {
            updateState('WAKE_LISTENING');
        } else {
            console.warn('Sleep word detected but not in LISTENING state');
        }
    },

    onSilenceDetected() {
        if (!this.handsFreeModeEnabled) {
            return;
        }
        if (this.state === 'LISTENING') {
            updateState('SENDING');
        }
    },

    activate() {
        if (this.isActive) throw new Error('Already active');
        this.isActive = true;
        if (this.handsFreeModeEnabled) {
            updateState('WAKE_LISTENING');
        } else {
            updateState('LISTENING');
        }
    },

    deactivate() {
        this.isActive = false;
        updateState('IDLE');
    },

    sendMessage() {
        if (!this.isActive) throw new Error('Not active');
        if (this.state === 'SENDING') throw new Error('Already sending');
        if (this.state === 'PLAYING') throw new Error('Cannot send while playing');
        if (this.state === 'LISTENING' || this.state === 'PAUSED') {
            updateState('SENDING');
        }
    },

    receiveResponse() {
        if (this.state !== 'SENDING') throw new Error('Not waiting for response');
        updateState('PLAYING');
    },

    finishPlayback() {
        if (this.state !== 'PLAYING') throw new Error('Not playing');
        updateState('LISTENING');
    },

    pauseListening() {
        if (this.state !== 'LISTENING') throw new Error('Cannot pause');
        updateState('PAUSED');
    },

    resumeListening() {
        if (this.state !== 'PAUSED') throw new Error('Not paused');
        updateState('LISTENING');
    },

    interruptPlayback() {
        if (this.state !== 'PLAYING') throw new Error('Not playing');
        updateState('LISTENING');
    },

    shouldBeListening() {
        return this.state === 'LISTENING' || this.state === 'WAKE_LISTENING';
    },

    shouldAutoPlay() {
        return this.isActive && this.state === 'SENDING';
    },

    shouldAutoRestart() {
        return this.isActive && this.state === 'PLAYING';
    }
};

// Dependencies injected from main dashboard
let VoiceInteraction = null;
let showNotification = null;
let StateIndicator = null;

/**
 * Initialize dependencies from main dashboard.
 * Must be called before using UI functions.
 */
export function initializeDependencies(deps) {
    VoiceInteraction = deps.VoiceInteraction;
    showNotification = deps.showNotification;
    StateIndicator = deps.StateIndicator;
}

/**
 * Toggle conversation mode on/off.
 */
export function toggleConversationMode() {
    if (conversationMode.isActive) {
        deactivateConversationMode();
    } else {
        activateConversationMode();
    }
}

/**
 * Activate conversation mode.
 * Starts auto-listening and enables hands-free loop.
 */
export function activateConversationMode() {
    try {
        conversationMode.activate();

        // Update button UI
        const btn = document.getElementById('conversation-mode-btn');
        btn.classList.add('active');

        // Keep mic button enabled for pause/resume functionality
        const voiceBtn = document.getElementById('voice-btn');
        voiceBtn.title = 'Pause/Resume Listening';

        // Auto-start listening
        VoiceInteraction.startListening();

        showNotification('Conversation Mode activated - speak naturally!', 'success');
    } catch (error) {
        console.error('Failed to activate conversation mode:', error);
        showNotification('Failed to activate conversation mode', 'error');
    }
}

/**
 * Deactivate conversation mode.
 * Stops listening/speaking and returns to manual mode.
 */
export function deactivateConversationMode() {
    conversationMode.deactivate();

    // Update button UI
    const btn = document.getElementById('conversation-mode-btn');
    btn.classList.remove('active');

    // Restore mic button title
    const voiceBtn = document.getElementById('voice-btn');
    voiceBtn.title = 'Voice Input';

    // Stop any active listening or speaking
    if (VoiceInteraction.getIsListening()) {
        VoiceInteraction.stopListening();
    }
    if (VoiceInteraction.getIsSpeaking()) {
        VoiceInteraction.cancelSpeech();
    }

    showNotification('Conversation Mode deactivated', 'info');
}

/**
 * Get the conversation mode state machine.
 * Used for dependency injection into other modules.
 *
 * @returns {Object} The conversation mode state machine
 */
export function getConversationMode() {
    return conversationMode;
}

/**
 * Get the wake word detector instance.
 * Used by voice_interaction for transcript processing.
 *
 * @returns {WakeWordDetector} The wake word detector instance
 */
export function getWakeWordDetector() {
    return wakeWordDetector;
}

/**
 * Get the silence detector instance.
 * Used by voice_interaction for silence checking.
 *
 * @returns {SilenceDetector} The silence detector instance
 */
export function getSilenceDetector() {
    return silenceDetector;
}

/**
 * Enable hands-free mode.
 * When enabled, conversation mode starts in WAKE_LISTENING state.
 */
export function enableHandsFreeMode() {
    conversationMode.enableHandsFreeMode();
    console.log('[ConversationMode] Hands-free mode enabled');
}

/**
 * Disable hands-free mode.
 * When disabled, conversation mode starts in LISTENING state.
 */
export function disableHandsFreeMode() {
    conversationMode.disableHandsFreeMode();
    console.log('[ConversationMode] Hands-free mode disabled');
}

File: src/prompt_manager/domain/conversation_mode.py
"""
Conversation Mode domain model.

Pure business logic for managing conversation mode state machine.
Orchestrates hands-free dialogue loop with zero dependencies.
"""
from enum import Enum


class ConversationState(Enum):
    """States in the conversation mode state machine.

    State Transitions:
    - IDLE: Not in conversation mode (default)
    - LISTENING: Actively listening for user speech
    - PAUSED: Conversation mode active but listening paused
    - SENDING: Sending message to LLM
    - PLAYING: Playing back LLM response
    """
    IDLE = "idle"
    LISTENING = "listening"
    PAUSED = "paused"
    SENDING = "sending"
    PLAYING = "playing"


class ConversationModeStateMachine:
    """State machine for conversation mode.

    Business Rules:
    - Starts in IDLE state (conversation mode off)
    - Activation starts listening automatically
    - Send message transitions from LISTENING/PAUSED to SENDING
    - Response receipt transitions from SENDING to PLAYING
    - Playback completion auto-restarts listening (loop)
    - Can pause/resume listening with mic button
    - Can interrupt playback to speak again
    - Deactivation returns to IDLE from any state
    """

    def __init__(self):
        """Initialize state machine in IDLE state."""
        self.current_state = ConversationState.IDLE
        self.is_active = False

    def activate(self) -> None:
        """Activate conversation mode and start listening.

        Business Rule: Automatically starts listening when activated.

        Raises:
            ValueError: If already active
        """
        if self.is_active:
            raise ValueError("Conversation mode already active")

        self.is_active = True
        self.current_state = ConversationState.LISTENING

    def deactivate(self) -> None:
        """Deactivate conversation mode and return to IDLE.

        Business Rule: Can deactivate from any state.
        """
        self.is_active = False
        self.current_state = ConversationState.IDLE

    def send_message(self) -> None:
        """Transition to SENDING state.

        Business Rules:
        - Can only send if conversation mode is active
        - Can send from LISTENING or PAUSED states
        - Cannot send if already sending or playing

        Raises:
            ValueError: If invalid state transition
        """
        if not self.is_active:
            raise ValueError("Conversation mode not active")

        if self.current_state == ConversationState.SENDING:
            raise ValueError("Already sending message")

        if self.current_state == ConversationState.PLAYING:
            raise ValueError("Cannot send while playing response")

        if self.current_state in (ConversationState.LISTENING, ConversationState.PAUSED):
            self.current_state = ConversationState.SENDING
        else:
            raise ValueError(f"Cannot send from state: {self.current_state}")

    def receive_response(self) -> None:
        """Transition to PLAYING state.

        Business Rule: Can only receive response if waiting for one (SENDING state).

        Raises:
            ValueError: If not in SENDING state
        """
        if self.current_state != ConversationState.SENDING:
            raise ValueError("Not waiting for response (not in SENDING state)")

        self.current_state = ConversationState.PLAYING

    def finish_playback(self) -> None:
        """Finish playing response and auto-restart listening.

        Business Rule: Auto-loops back to LISTENING after playback completes.

        Raises:
            ValueError: If not in PLAYING state
        """
        if self.current_state != ConversationState.PLAYING:
            raise ValueError("Not currently playing response")

        self.current_state = ConversationState.LISTENING

    def pause_listening(self) -> None:
        """Pause listening (user clicks mic button).

        Business Rules:
        - Can only pause from LISTENING state
        - Cannot pause during SENDING or PLAYING

        Raises:
            ValueError: If invalid state transition
        """
        if self.current_state != ConversationState.LISTENING:
            raise ValueError(f"Cannot pause from state: {self.current_state}")

        self.current_state = ConversationState.PAUSED

    def resume_listening(self) -> None:
        """Resume listening from paused state.

        Raises:
            ValueError: If not in PAUSED state
        """
        if self.current_state != ConversationState.PAUSED:
            raise ValueError("Not in paused state")

        self.current_state = ConversationState.LISTENING

    def interrupt_playback(self) -> None:
        """Interrupt playback and return to listening.

        Business Rule: User can interrupt playback to speak again.

        Raises:
            ValueError: If not in PLAYING state
        """
        if self.current_state != ConversationState.PLAYING:
            raise ValueError("Not currently playing response")

        self.current_state = ConversationState.LISTENING

    def should_be_listening(self) -> bool:
        """Query if microphone should be active.

        Returns:
            True if in LISTENING state, False otherwise
        """
        return self.current_state == ConversationState.LISTENING

    def should_auto_play(self) -> bool:
        """Query if response should auto-play.

        Business Rule: Auto-play only when in conversation mode.

        Returns:
            True if in conversation mode and waiting for response
        """
        return self.is_active and self.current_state == ConversationState.SENDING

    def should_auto_restart(self) -> bool:
        """Query if should auto-restart listening after playback.

        Business Rule: Auto-restart only when in conversation mode.

        Returns:
            True if in conversation mode and currently playing
        """
        return self.is_active and self.current_state == ConversationState.PLAYING
